sssd_test_framework/utils/sshd.py

```diff
@@ -0,0 +1,128 @@
+"""SSH Daemon Tools."""
+
+from __future__ import annotations
+
+from pytest_mh import MultihostHost, MultihostUtility
+from pytest_mh.ssh import SSHProcessResult
+from pytest_mh.utils.fs import LinuxFileSystem
+from pytest_mh.utils.services import SystemdServices
+
+__all__ = [
+    "SSHDUtils",
+]
+
+
+class SSHDUtils(MultihostUtility[MultihostHost]):
+    """
+    Managing global and server SSH configuration files.
+
+    .. code-block:: python
+        :caption: Example usage
+
+        @pytest.mark.topology(KnownTopologyGroup.AnyProvider)
+        def test_example(client: Client, provider: GenericProvider):
+            # Add user
+            provider.user("user1").add()
+
+            # Select authselect profile and feature
+            client.authselect.select("sssd", ["with-gssapi", "with-mkhomedir"])
+
+            # Start sssd
+            client.sssd.start()
+
+            # Configure ssh client and daemon
+            client.sshd.config_set([{"GSSAPIAuthentication": "yes"}])
+
+            # Reload sshd configuration
+            client.sshd.reload()
+    """
+
+    def __init__(
+        self, host: MultihostHost, fs: LinuxFileSystem, svc: SystemdServices, file: str = "/etc/ssh/sshd_config"
+    ) -> None:
+        """
+        :param host: Remote host instance.
+        :type host: MultihostHost
+        :param fs: Linux file system instance.
+        :type fs: LinuxFileSystem
+        :param svc: Systemd utils.
+        :type svc: SystemdServices
+        :param file: Configuration file.
+        :type file: str
+        """
+        super().__init__(host)
+        self.__backup: str | None
+        self.fs: LinuxFileSystem = fs
+        self.svc: SystemdServices = svc
+
+        self.file: str = file
+        self.path: str = "/files" + self.file
+        self.args: str = f'--transform "sshd.lns incl {self.file}"'
+        self.cmd: str = ""
+
+    def setup_when_used(self) -> None:
+        super().setup_when_used()
+        self.fs.backup(self.file)
```

Comment:
I was about to push it, but this also requires a teardown to reload original configuration. Something like:

```
def teardown_when_used(self) -> None:
    self.fs.restore(self.file)
    self.svc.reload("sshd")
```

Reply:
I thought pytest-mh , when using self.fs.backup restored the file. This needs to be updated for the Pam util class I wrote to then.

Reply:
Moving to draft, something isn't right.

Reply:
Okay, self.fs.restore(self.file) isn't necessary, the file does get restored.

Reply:
https://pastebin.com/TiNs0Q3S

I put a print after the config_set and another in the teardown, the changes have already been reverted.

Reply:
I don't for see many tests using this function at all, this is mainly to enable GSSAPIAuthentication on the host which can be done safely with little impact. I updated the comment with a warning when using this. There're only a few tests that need to edit the sshd config.

Reply:
Yes, at least for the moment.



> Okay, self.fs.restore(self.file) isn't necessary, the file does get restored.

This is something we need to work on in pytest-mh. To build a dependency tree so it is clear in which order are the MultihostUtility objects destroyed. But it works for us at the moment.